File: apps/apple/TeslaSync/feature-views/TOTPEnrollmentSection.Source.swift
//
//  TOTPEnrollmentSection.Source.swift
//  TeslaSync — P4 feature view · 0217 · TOTPEnrollmentSection (Apple)
//
//  The in-memory `TOTPEnrollmentSource` that previews + unit tests drive the
//  surface with (the production app supplies a real implementation over the
//  shared P1/S8 state holders). Seeds an optional initial status snapshot on
//  `start()`, lets a test push further snapshots via `push(_:)`, and returns
//  caller-configured mutation outcomes while recording call counts. No network,
//  no real store. Also defines the deterministic sample enrollment used by both.
//

import Foundation

// MARK: - In-memory source (previews + tests)

/// In-memory source for previews + unit tests. Seeds an optional initial status
/// snapshot on `start()`, lets a test push further snapshots via `push(_:)`, and
/// returns caller-configured mutation outcomes while recording call counts. No
/// network, no real store.
@MainActor
public final class InMemoryTOTPEnrollmentSource: TOTPEnrollmentSource {
    public var onUpdate: (@MainActor (TOTPEnrollmentUpdate) -> Void)?
    public private(set) var startCount = 0
    public private(set) var stopCount = 0
    public private(set) var refreshCount = 0
    public private(set) var enrollCount = 0
    public private(set) var verifiedCodes: [String] = []
    public private(set) var revokeCount = 0
    public private(set) var regenerateCount = 0

    /// The enrollment returned by `enroll()` (web mutation result).
    public var enrollResult: TOTPEnrollmentData
    /// When non-nil, `verify(code:)` throws it; otherwise it succeeds.
    public var verifyFailure: TOTPVerifyError?
    /// The codes returned by `regenerateBackupCodes()`.
    public var regenerateResult: [String]
    /// When non-nil, `enroll()` / `revoke()` / `regenerate()` throw it.
    public var actionFailure: Error?

    private let initial: TOTPEnrollmentUpdate?

    public init(
        initial: TOTPEnrollmentUpdate? = nil,
        enrollResult: TOTPEnrollmentData = .preview,
        verifyFailure: TOTPVerifyError? = nil,
        regenerateResult: [String] = ["regen-1111", "regen-2222"],
        actionFailure: Error? = nil
    ) {
        self.initial = initial
        self.enrollResult = enrollResult
        self.verifyFailure = verifyFailure
        self.regenerateResult = regenerateResult
        self.actionFailure = actionFailure
    }

    public func start() {
        startCount += 1
        if let initial { onUpdate?(initial) }
    }

    public func stop() {
        stopCount += 1
    }

    public func refresh() {
        refreshCount += 1
    }

    public func enroll() async throws -> TOTPEnrollmentData {
        enrollCount += 1
        if let actionFailure { throw actionFailure }
        return enrollResult
    }

    public func verify(code: String) async throws {
        verifiedCodes.append(code)
        if let verifyFailure { throw verifyFailure }
    }

    public func revoke() async throws {
        revokeCount += 1
        if let actionFailure { throw actionFailure }
    }

    public func regenerateBackupCodes() async throws -> [String] {
        regenerateCount += 1
        if let actionFailure { throw actionFailure }
        return regenerateResult
    }

    /// Pushes a snapshot to the bound model (test / preview affordance).
    public func push(_ update: TOTPEnrollmentUpdate) {
        onUpdate?(update)
    }
}

// MARK: - Sample enrollment

public extension TOTPEnrollmentData {
    /// A deterministic sample enrollment for previews + the in-memory source.
    static let preview = TOTPEnrollmentData(
        secret: "JBSWY3DPEHPK3PXP",
        otpauthURI: "otpauth://totp/TeslaSync:owner?secret=JBSWY3DPEHPK3PXP&issuer=TeslaSync",
        qrDataURI: "data:image/png;base64,iVBORw0KGgo=",
        backupCodes: [
            "11aa-22bb", "33cc-44dd", "55ee-66ff",
            "77gg-88hh", "99ii-00jj", "abcd-efgh"
        ],
        expiresAt: nil
    )
}

File: apps/android/app/src/main/java/com/teslasync/feature-views/TOTPEnrollmentSection/TOTPEnrollmentSectionSource.kt
// The data port the TOTPEnrollmentSection feature view binds to — the native analogue of the `useTOTP` hook
// domain the web component composes (web/src/api/hooks/useTOTP.ts → web/src/features/settings/components/
// TOTPEnrollmentSection.tsx): the cache-then-network status read plus the enroll / verify / revoke /
// regenerate-backup-codes mutations, each of which refreshes the status feed on success exactly as the web
// hooks invalidate `totpKeys.status`. The view performs NO HTTP itself; a shared adapter (the S8 TOTPStore or
// the S7 TOTPRepository) or a test/preview fake drives this seam. Cache-then-network freshness is preserved
// end to end (ADR-013): every status emission's cached/stale/error flags flow through unchanged so the
// view-model can render the full state matrix.
//
// `InvalidPackageDeclaration`/`ktlint:standard:filename`/`MatchingDeclarationName` are suppressed: the mandated
// surface directory (com/teslasync/feature-views/TOTPEnrollmentSection) cannot form a valid Kotlin package and
// the file hosts the seam plus its bindings + a preview/test fake, mirroring the sibling surfaces.
@file:Suppress("InvalidPackageDeclaration", "ktlint:standard:filename", "MatchingDeclarationName")

package io.teslasync.android.featureviews.totpenrollmentsection

import io.teslasync.shared.core.data.repo.Resource
import io.teslasync.shared.core.data.repo.TOTPRepository
import io.teslasync.shared.core.presentation.totp.TOTPBackupCodesResponse
import io.teslasync.shared.core.presentation.totp.TOTPEnrollment
import io.teslasync.shared.core.presentation.totp.TOTPStatus
import io.teslasync.shared.core.presentation.totp.TOTPStore
import io.teslasync.shared.core.presentation.totp.TOTPVerifyResult
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * The single seam the [TOTPEnrollmentSectionViewModel] depends on so it binds to an abstraction (real adapter
 * ↔ test fake), never to a concrete store/repository or the network. [status] is the cache-then-network feed
 * the web `useTOTPStatus` hook serves; [enroll]/[verify]/[revoke]/[regenerateBackupCodes] mirror the web
 * mutations (each non-throwing, refreshing the status feed on success); [refresh] re-runs the status read
 * backing the surface's retry affordance. No HTTP touches the view.
 */
interface TOTPEnrollmentSectionSource {
    /** Stream the cache-then-network per-user TOTP status (web `useTOTPStatus`, `GET /auth/totp`). */
    fun status(): Flow<Resource<TOTPStatus>>

    /** Start a fresh enrollment, returning the secret/QR/one-time codes (web `useTOTPEnroll`). */
    suspend fun enroll(): Result<TOTPEnrollment>

    /** Promote a pending enrollment to active with [code] (web `useTOTPVerify`). */
    suspend fun verify(code: String): Result<TOTPVerifyResult>

    /** Disable TOTP for the subject (web `useTOTPRevoke`). */
    suspend fun revoke(): Result<Unit>

    /** Rotate the backup-code set, returned once (web `useTOTPRegenerateBackupCodes`). */
    suspend fun regenerateBackupCodes(): Result<TOTPBackupCodesResponse>

    /** Re-run the status read (web `refetch()` / the hooks' `invalidateAndBroadcast`). */
    fun refresh()
}

/**
 * Binds the surface to the shared **S8** [TOTPStore] — the memoized, multi-observer holder every two-factor
 * surface shares app-wide. The store's hot status feed is re-collected through the view-model's refresh
 * trigger, and each mutation already refreshes that feed on success (the web invalidate-on-mutation rule), so
 * the section flips from "Not enrolled" to "Active" without a manual refetch. No HTTP touches the view.
 */
fun bindTOTPEnrollmentSectionSource(store: TOTPStore): TOTPEnrollmentSectionSource =
    object : TOTPEnrollmentSectionSource {
        override fun status(): Flow<Resource<TOTPStatus>> = store.status

        override suspend fun enroll(): Result<TOTPEnrollment> = store.enroll()

        override suspend fun verify(code: String): Result<TOTPVerifyResult> = store.verify(code)

        override suspend fun revoke(): Result<Unit> = store.revoke()

        override suspend fun regenerateBackupCodes(): Result<TOTPBackupCodesResponse> = store.regenerateBackupCodes()

        override fun refresh() = store.refresh()
    }

/**
 * Binds the surface directly to the shared **S7** [TOTPRepository]. Each [status] call starts a NEW
 * cache-then-network collection, so the view-model's refresh/retry trigger a genuine re-fetch (the web
 * `refetch()` behaviour); the repository evicts the status cache key on every mutation success exactly as the
 * web hooks invalidate `totpKeys.status`, so the post-mutation re-collection reflects the new state. Use this
 * binding when a host does not share a single app-wide store.
 */
fun bindTOTPEnrollmentSectionSource(repository: TOTPRepository): TOTPEnrollmentSectionSource =
    object : TOTPEnrollmentSectionSource {
        override fun status(): Flow<Resource<TOTPStatus>> = repository.status()

        override suspend fun enroll(): Result<TOTPEnrollment> = repository.enroll()

        override suspend fun verify(code: String): Result<TOTPVerifyResult> = repository.verify(code)

        override suspend fun revoke(): Result<Unit> = repository.revoke()

        override suspend fun regenerateBackupCodes(): Result<TOTPBackupCodesResponse> = repository.regenerateBackupCodes()

        override fun refresh() = Unit
    }

/**
 * An in-memory [TOTPEnrollmentSectionSource] for `@Preview`s and tests. It replays a mutable [status] value
 * (settable via [emit] to walk the surface through its states) and returns the canned mutation [Result]s,
 * counting each call so a test can assert the view-model invoked the right one. It performs no networking and
 * is the default a host injects when no shared store is wired (e.g. a preview harness).
 */
class InMemoryTOTPEnrollmentSectionSource(
    initial: Resource<TOTPStatus> = Resource.Loading(cached = null, fetchedAt = null, stale = false),
    private val enrollResult: Result<TOTPEnrollment> = Result.failure(IllegalStateException("no enroll configured")),
    private val verifyResult: Result<TOTPVerifyResult> = Result.success(TOTPVerifyResult(activated = true)),
    private val revokeResult: Result<Unit> = Result.success(Unit),
    private val regenerateResult: Result<TOTPBackupCodesResponse> = Result.success(TOTPBackupCodesResponse()),
) : TOTPEnrollmentSectionSource {
    private val statusState = MutableStateFlow(initial)

    /** Call counters a test asserts against to prove the view-model routed to the right mutation. */
    var enrollCalls: Int = 0
        private set
    var verifyCalls: Int = 0
        private set
    var revokeCalls: Int = 0
        private set
    var regenerateCalls: Int = 0
        private set
    var refreshCalls: Int = 0
        private set

    /** The last code passed to [verify]; lets a test confirm the sanitised code reached the seam. */
    var lastVerifiedCode: String? = null
        private set

    /** Pushes a new status [resource] so a test/preview can walk the surface through its state matrix. */
    fun emit(resource: Resource<TOTPStatus>) = statusState.update { resource }

    override fun status(): Flow<Resource<TOTPStatus>> = statusState.asStateFlow()

    override suspend fun enroll(): Result<TOTPEnrollment> {
        enrollCalls++
        return enrollResult
    }

    override suspend fun verify(code: String): Result<TOTPVerifyResult> {
        verifyCalls++
        lastVerifiedCode = code
        return verifyResult
    }

    override suspend fun revoke(): Result<Unit> {
        revokeCalls++
        return revokeResult
    }

    override suspend fun regenerateBackupCodes(): Result<TOTPBackupCodesResponse> {
        regenerateCalls++
        return regenerateResult
    }

    override fun refresh() {
        refreshCalls++
    }
}
